### ISSUE TITLE
[iOS 10] Cannot reset ads on iOS 

### ISSUE THREAD

**Issue by stoicbuddha** [googleads/videojs-ima#426]
We utilize Iris.tv for one our clients.  It allows a user to skip to a new video (provided by their api in chunks of 5 videos), like or dislike a video, etc.

When a video switch happens, the video source is changed, and we do another call to ads.  On Desktop and Android, this works perfectly fine.  On iOS 10 (Safari), this fails.  It draws a black div and hides the player, and does nothing.  It makes no network calls to any ad server, and doesn't appear to do anything.  I have an ad events plugin that acknowledges a timeout (a custom timeout on our end), so I know the ads plugin is triggering.

Is there a way to switch out ads in iOS that maybe I'm not seeing?  I currently just call `player.ima(options)` again and that works on Android and iOS.

Player: http://staging.player.dotstudiopro.com/share/58c48e8397f8158a5e1ad4ff?targetelm=.player&companykey=56798e9197f815d36f339ee4&loopplayback=false&autostart=true&muteonstart=false&enablesharing=false&skin=0000ff

**Comment by shawnbuso** [comment]
Hey there,

I think our [playlist example](https://github.com/googleads/videojs-ima/tree/master/examples/playlist) is exactly what you're looking for. That example prevents users from changing the video while the ad is playing, so you'll need to wait until you see the content video to try to switch. Is that working on the devices you're using? If not, could you let me know which device and iOS versions you're using and at what point you're trying to switch videos?

Thanks,
Shawn

**Comment by chrisboustead** [comment]
It seems (to me at least) that the once **one** video ends, it is not possible to successfully play another video w/ads.

I modified the playlist example, to call another video play w/ adTag on the `ended` event: https://gist.github.com/chrisboustead/a5af07fc72c4bd62f594238a37a3b633

```
    // On 'ended' event of first video:
    player.ima.setContentWithAdTag(
          'some valid source',
          'some valid adtag',
    true);
    player.ima.requestAds();
```

**Error:** 

```
TypeError: Cannot read property 'trackChangeHandler' of null
    at Player.<anonymous> (videojs.ads.js:202)
    at HTMLDivElement.bound (video.js:2232)
    at HTMLDivElement.data.dispatcher (video.js:1867)
    at trigger (video.js:1994)
    at Player.trigger$$1 [as trigger] (video.js:2795)
    at Player.checkSrc (videojs.ads.js:481)
    at HTMLDivElement.bound (video.js:2232)
    at HTMLDivElement.data.dispatcher (video.js:1867)
    at trigger (video.js:1994)
    at Player.trigger$$1 [as trigger] (video.js:2795)
```

**More Detail:**
Seems that `reset` is called before `player.ima.setContentWithAdTag` is, and `snapshot` (among other properties) are no longer objects, needing re-initialization. 
```
function initialize(player) {
  player.on('contentupdate', function () {
>>> if (player.ads.snapshot.trackChangeHandler) {
      var textTrackList = player.textTracks();

      textTrackList.removeEventListener('change', player.ads.snapshot.trackChangeHandler);
      player.ads.snapshotIOSTrackHandlerAdded_ = false;
    }
  });

  player.ads.snapshotInitialized_ = true;
}
```

**Dependencies:**
- videojs v6.2.8
- videojs-contrib-ads v5.1.0
- videojs-ima v0.6.0

**Comment by shawnbuso** [comment]
Are you seeing this issue with older video.js versions? [Currently we only support up to 5.3](https://github.com/googleads/videojs-ima/blob/master/package.json#L27).

Also, can you try requesting more ads in a [contentAndAdsEndedListener](https://github.com/googleads/videojs-ima/blob/master/docs/api/this.md#addcontentandadsendedlistener-listener-) instead of on the videoPlayer's ended event? That will avoid breaking any state we maintain in the plugin. Thanks!

**Comment by chrisboustead** [comment]
Sorry for the confusion with versions there, tested with:

- videojs 5.20.3
- videojs-ima 0.7.0
- videojs-contrib-ads 5.0.3

Tested and updated the gist with the contentAndAdsEndedListener, results are similar.  No pre-roll after the first video.  Interestingly enough, if you seek or pause/play the second video the pre-roll will play, and content will resume once ad is complete.  Probably unrelated to the initial issue, can open an separate issue to track.

**Comment by chrisboustead** [comment]
Actually, even the playlist example right out of the box (only updating the paths for videojs-contrib-ads assets) no longer works even when clicking on the second video.  

Only the ads for the first video are played.  If there is interaction after clicking the second video (seek/pause -> play) then the ads for the second video will kick off.

**Comment by shawnbuso** [comment]
I just noticed that too - that's indeed a bug. We'll take a look and see what we need to do to get it remedied.

**Comment by chrisboustead** [comment]
Hey @shawnbuso, I've looked into this a little but haven't been successful finding a resolution.  Have you made any progress or do you have any thoughts about what I should look into?

I was able to have some success using IMA SDK directly with videoJS, without this plugin or the contrib-ads plugin (html5 only), but is far from a production-ready solution.

**Comment by shawnbuso** [comment]
Hey sorry about this, I've actually been working with the contrib-ads folks on troubleshooting this one. You can follow our progress in [this issue](https://github.com/videojs/videojs-contrib-ads/issues/291).

**Comment by chrisboustead** [comment]
Appreciate it!  Jumping over there...

**Comment by shawnbuso** [comment]
It looks like this is fixed with contrib-ads 5.1.3. The plugin in its current state isn't yet updated to use that version, but I'm working on a refactor that will get the plugin up to date with the latest contrib-ads. Once that refactor is complete we should be good to go. If you want to take a crack at using 5.1.3 with the plugin in its current form you can give it a shot and let me know if it works.

**Comment by chrisboustead** [comment]
Confirmed- works with 5.1.3.  

Not sure yet if there's any clashes with the upgrade but simple pre-rolls worked with:

videojs: 5.2.8
videojs-contrib-ads: 5.1.3
videojs-ima: 0.6.0

As far as i'm concerned this issue can be closed as it was regarding subsequent prerolls only, but that's your call.  Thanks for the help @shawnbuso!

**Comment by shawnbuso** [comment]
Sweet, glad it's working!
